medmodels/data_synthesis/synthesizer.py

```diff
@@ -0,0 +1,132 @@
+"""Abstract class for synthesizers."""
+
+from __future__ import annotations
+
+from abc import ABCMeta, abstractmethod
+from pathlib import Path
+from typing import TYPE_CHECKING, Any, Literal, Optional, Type
+
+import torch
+from torch import nn
+
+from medmodels.data_synthesis.synthesizer_model import SynthesizerModel
+from medmodels.medrecord.medrecord import MedRecord
+
+if TYPE_CHECKING:
+    from medmodels.data_synthesis.builder import SynthesizerBuilder
+
+
+class Synthesizer(nn.Module, metaclass=ABCMeta):
+    """Synthesizer is the abstract class to define synthesizers."""
+
+    _preprocessor: Type[nn.Module]
+    _postprocessor: Type[nn.Module]
+    device: torch.device
+
+    def __init__(
+        self,
+        preprocessor: Type[nn.Module],
+        postprocessor: Type[nn.Module],
+        device: Literal["cpu", "cuda"] = "cuda",
```

Comment:
there are many more types other than `cpu` or `cupa`, including e.g. `mps` for macbooks.

I'd probably suggest
```suggestion
        device: torch.device = torch.device("cpu"),
```